test/e2e/keys.e2e.js

```diff
@@ -47,6 +69,148 @@ describe('Keys Commands [@device]', () => {
+			expect(exitCode).to.equal(1);
+		});
+	});
+
+	describe('Save Subcommand', () => {
+		const filename = path.join(PATH_TMP_DIR, `${DEVICE_NAME}.pem`);
+		const expectedKeys = [`${DEVICE_NAME}.der`, `${DEVICE_NAME}.pub.pem`];
+		const help = [
+			'Save a key from your device to a file',
+			'Usage: particle keys save [options] <filename>',
+			'',
+			'Global Options:',
+			'  -v, --verbose  Increases how much logging to display  [count]',
+			'  -q, --quiet    Decreases how much logging to display  [count]',
+			'',
+			'Options:',
+			'  --force  Force overwriting of <filename> if it exists  [boolean] [default: false]'
+		];
+
+		afterEach(async () => {
+			await cli.resetDevice();
+		});
+
+		after(async () => {
+			// TODO (mirande): replace w/ `cli.waitForDeviceToGetOnline()`
+			// helper when available
+			await delay(30 * 1000);
+		});
+
+		it('Saves device keys', async () => {
+			await cli.enterDFUMode();
+			const { stdout, stderr, exitCode } = await cli.run(['keys', 'save', filename]);
+
+			expect(stdout).to.equal('Saved!');
+			expect(stderr).to.equal('');
+			expect(exitCode).to.equal(0);
+			for (const key of expectedKeys){
+				const stats = await fs.stat(path.join(PATH_TMP_DIR, key));
+				expect(stats.size).to.be.above(100);
+			}
+		});
+
+		it('Fails to save device keys when device is not in DFU mode', async () => {
+			const { stdout, stderr, exitCode } = await cli.run(['keys', 'save', filename]);
+
+			expect(stdout.split('\n')).to.include.members(dfuInstructions);
+			expect(stderr).to.equal('');
+			expect(exitCode).to.equal(0); // TODO (mirande): fix: inconsistent error code - see `particle new`
+		});
+
+		it('Fails to save device keys when filename is not provided', async () => {
+			const { stdout, stderr, exitCode } = await cli.run(['keys', 'save']);
+
+			expect(stdout).to.equal('Parameter \'filename\' is required.');
+			expect(stderr.split('\n')).to.include.members(help);
+			expect(exitCode).to.equal(1);
+		});
+	});
+
+	describe('Server Subcommand', () => {
+		before(async () => {
+			await cli.setTestProfileAndLogin();
+			await cli.flashStrobyFirmwareOTAForTest();
+		});
+
+		after(async () => {
+			await cli.logout();
+			await cli.setDefaultProfile();
+		});
+
+		it('Switches server public keys', async () => {
+			await cli.enterDFUMode();
+			const { stdout, stderr, exitCode } = await cli.run(['keys', 'server']);
+
+			expect(stdout).to.equal('Okay!  New keys in place, your device will not restart.');
+			expect(stderr).to.equal('');
+			expect(exitCode).to.equal(0);
+
+			await cli.resetDevice();
+			await cli.waitForVariable('name', 'stroby');
+		}).timeout(5 * 60 * 1000);
+
+		it('Fails to switch server public keys when device is not in DFU mode', async () => {
+			const { stdout, stderr, exitCode } = await cli.run(['keys', 'server']);
+
+			expect(stdout.split('\n')).to.include.members(dfuInstructions);
+			expect(stderr).to.equal('');
+			expect(exitCode).to.equal(1);
+		});
+	});
```

Comment:
@JamesHagerman added these so at least we have a examples that you can build on and some supporting tooling but they're slower and flakier than i'd like - mostly due to the time it takes for a device to connect :(

one thing it did highlight is that the `particle keys server --deviceType photon` command form doesn't appear to do anything. it should probably be a usage error ([example](https://github.com/particle-iot/particle-cli/commit/6cbb33eab9c308cbc5a4135834eb2a3a808cb443#diff-1c5e41ba62634f0609f01bdf97c1a93fR155))..?